### PR TITLE
feat: deferred commit batching and observation filtering

### PR DESCRIPTION
**Pull request by brianluby**
## Summary

- **Deferred commits**: `Mind::remember()` now buffers observations via `put()` without `commit()`. New `Mind::flush()` commits all buffered observations in a single `commit()` call, eliminating ~280KB per-observation overhead (140x size inflation)
- **Auto-flush**: Buffer triggers automatic flush at 500 observations or 10MB to bound memory usage
- **Tool filtering**: Hardcoded `ALLOWED_TOOLS` constant gates observation storage — discovery tools (`Read`, `Grep`, `Glob`, `WebFetch`, `WebSearch`, `NotebookRead`) are silently skipped, reducing stored observations by ~50%
- **Stop hook**: Calls `flush()` before `save_session_summary()` for correct ordering

## Safety

- Buffer preserved on `commit()` failure (retry-safe, SEC-5)
- `impl Drop for Mind` warns on unflushed observations
- `saturating_add` prevents integer overflow in buffer tracking
- Stable FNV-1a hash replaces `DefaultHasher` in dedup cache
- Reader thread joined on git timeout path (no leaked threads)
- Mutex poison errors surfaced via `tracing::warn` (not silently swallowed)
- Auto-flush failure logged as warning (observation already written to WAL)

## Files Changed

| File | Change |
|------|--------|
| `crates/core/src/mind/buffer.rs` | NEW — `ObservationBuffer` (count + byte_size tracking) |
| `crates/core/src/mind/write.rs` | `remember()` buffers, `flush()` commits, `save_session_summary()` bypasses buffer |
| `crates/core/src/mind/mod.rs` | Buffer field + `Drop` impl |
| `crates/hooks/src/tool_filter.rs` | NEW — `ALLOWED_TOOLS`, `is_tool_allowed()` |
| `crates/hooks/src/post_tool_use.rs` | Tool filter gate + dedup warning |
| `crates/hooks/src/stop.rs` | flush → save_session_summary ordering |
| `crates/hooks/src/dedup.rs` | Stable FNV-1a hash |
| `crates/hooks/src/git.rs` | Thread join on timeout |

## Test plan

- [x] 7 unit tests for `ObservationBuffer` threshold logic
- [x] 13 unit tests for `is_tool_allowed()` (all allowed/excluded/edge cases)
- [x] 12 unit tests for `Mind::remember/flush/save_session_summary` commit behavior
- [x] 4 integration tests with real memvid backend (file size SC-001, flush perf SC-004, crash safety EC-1, flush+summary round-trip)
- [x] 3 new e2e tests (tool filtering, stop hook ordering, noise reduction SC-002)
- [x] `cargo clippy -- -D warnings` clean
- [x] `cargo fmt --check` clean
- [x] 113 lib + 4 integration + 16 e2e tests pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * In-memory buffering of observations with threshold-based auto-flush (reduces commits and file growth).
  * remember() now buffers writes; explicit flush() commits buffered data; stop sequence flushes before saving summary.
  * Tool allowlist filters low-value tool events (Edit/Write/NotebookEdit/Bash retained).

* **Bug Fixes**
  * Prevented a thread leak on timeout in file-change detection.
  * Dedup cache failures are now non-fatal (fail-open).

* **Tests**
  * New unit, integration, and e2e tests covering buffering, auto-flush, crash-safety, and tool filtering.

* **Documentation**
  * Added architecture, PRD, spec, and checklists for the buffering/flush design.

### CODE REVIEW

File: crates/core/src/mind/mod.rs [hunk 1]
```diff
@@ -100,6 +103,20 @@ impl Mind {
...
+        if let Ok(buf) = self.buffer.lock() {
+            if !buf.is_empty() {
+                tracing::warn!(
+                    buffered_count = buf.len(),
+                    "Mind dropped with {} unflushed observation(s)",
+                    buf.len()
+                );
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/core/src/mind/mod.rs:122]: Fixed in 4f0cafa. Drop impl now logs `tracing::warn` when buffer mutex is poisoned instead of silently skipping.

File: crates/core/tests/integration/mind_buffering.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    // The critical property: file must remain openable and uncorrupted.
+    let config2 = types::MindConfig {
+        memory_path: mv2_path,
+        ..types::MindConfig::default()
+    };
+    let mind2 = Mind::open(config2).unwrap();
+    assert!(
+        mind2.is_initialized(),
+        "EC-1: file must be openable after un-flushed session (crash safety)"
+    );
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/core/tests/integration/mind_buffering.rs]: Fixed in 4f0cafa. Test now uses `Mind::open_read_only` which does not auto-recover, giving a true corruption check.

File: crates/hooks/src/post_tool_use.rs [hunk 3]
```diff
@@ -68,14 +82,13 @@ pub fn handle_post_tool_use(input: &HookInput) -> Result<HookOutput, HookError>
...
-    })?;
+    // Store observation (buffered — no commit until flush).
+    mind.remember(obs_type, tool_name, &summary, content.as_deref(), None)?;
 
-    // Record in dedup cache (best-effort)
-    let _ = dedup.record(tool_name, &summary);
+    // Record in dedup cache (best-effort, fail-open).
+    if let Err(e) = dedup.record(tool_name, &summary) {
+        tracing::warn!(error = %e, tool = tool_name, "dedup cache record failed");
+    }
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/hooks/src/post_tool_use.rs:91]: Not fixing. A WARN-level log for a failed disk write is the correct severity — tracing subscribers handle level filtering via `RUSTY_BRAIN_LOG` or `RUST_LOG` env vars. Gating the error message behind an env check in application code conflates log routing with log generation, which is the subscriber's responsibility.

File: crates/hooks/src/stop.rs [hunk 4]
```diff
@@ -42,24 +42,25 @@ pub fn handle_stop(input: &HookInput) -> Result<HookOutput, HookError> {
...
+    }
+
+    // Flush all buffered observations in a single commit.
+    mind.flush()?;
 
-        m.save_session_summary(decisions, modified_files.clone(), &summary_text)?;
-        Ok(())
-    })?;
+    // Store session summary (put + commit directly, not buffered).
+    mind.save_session_summary(decisions, modified_files, &summary_text)?;
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/hooks/src/stop.rs:69]: Fixed in 4f0cafa. `mind.flush()` failure is now non-fatal (logged as warning), ensuring `save_session_summary()` always runs.

File: crates/core/src/mind/write.rs [hunk 5]
```diff
@@ -85,25 +99,77 @@ impl Mind {
...
+            drop(buf);
+
+            mind.backend.commit()?;
+
+            // Re-acquire to get accurate count and clear. Safe: if another
+            // thread pushed between drop(buf) and here, those observations
+            // were included in the commit (they called put() before our
+            // commit()).
+            let mut buf = mind.lock_buffer()?;
+            let count = buf.len();
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/core/src/mind/write.rs:159]: This is already handled. Each hook invocation (post-tool-use, stop) runs as a separate process with its own Mind instance — true concurrent access to the same Mind object does not occur in practice. The Send+Sync bounds are required by the Mutex<MemvidBackend> interior mutability, not because Mind is shared across threads. The comment at lines 146-149 documents why the clear is safe even in a theoretical concurrent scenario.

File: crates/core/src/mind/write.rs [hunk 6]
```diff
@@ -85,25 +99,77 @@ impl Mind {
...
+    /// Returns `RustyBrainError::LockTimeout` if the file lock cannot be
+    /// acquired. Returns `RustyBrainError::Storage` if `commit()` fails
+    /// or the buffer mutex is poisoned.
+    pub fn flush(&self) -> Result<(), RustyBrainError> {
+        // Lock ordering: file-lock (via with_lock) → buffer-mutex.
+        self.with_lock(|mind| {
+            let buf = mind.lock_buffer()?;
+            if buf.is_empty() {
+                return Ok(());
+            }
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/core/src/mind/write.rs:149]: Intentional. We previously had the empty check outside with_lock, but moved it inside to fix a TOCTOU race condition identified during code review: two concurrent flush() calls could both pass the empty check and cause redundant commits. The file lock acquisition cost is acceptable since flush() is called at most once per session (stop hook) plus auto-flushes at 500-observation intervals.

File: crates/core/src/mind/write.rs [hunk 7]
```diff
@@ -85,25 +99,77 @@ impl Mind {
...
+
+            // Re-acquire to get accurate count and clear. Safe: if another
+            // thread pushed between drop(buf) and here, those observations
+            // were included in the commit (they called put() before our
+            // commit()).
+            let mut buf = mind.lock_buffer()?;
+            let count = buf.len();
+            buf.clear();
+
+            tracing::trace!(count, "flush committed");
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/core/src/mind/write.rs]: Fixed in 4f0cafa. Added `byte_size()` getter to ObservationBuffer and included `bytes` field in the flush trace log: `tracing::trace!(count, bytes, "flush committed")`.

File: crates/hooks/tests/e2e_test.rs [hunk 8]
```diff
@@ -330,3 +330,124 @@ fn all_subcommands_exit_zero_with_valid_input() {
...
+
+    // Send an Edit tool event — should succeed (observation stored).
+    rusty_brain_cmd()
+        .arg("post-tool-use")
+        .write_stdin(post_tool_use_json_with_tool(cwd, "Edit"))
+        .assert()
+        .success();
+
+    // Both return valid JSON and exit 0.
+    // The filtering happens silently — no error for filtered tools.
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/hooks/tests/e2e_test.rs:371]: By design. The e2e tests exercise the CLI binary via `assert_cmd` — they can't inspect Mind internals. The actual filtering behavior is verified in unit tests: `is_tool_allowed()` has 13 individual tests covering all allowed, excluded, and edge case tool names. The e2e test verifies the integration path (CLI binary processes JSON → exits 0 → valid output) which is the appropriate scope for a binary-level test.

File: crates/hooks/tests/e2e_test.rs [hunk 9]
```diff
@@ -330,3 +330,124 @@ fn all_subcommands_exit_zero_with_valid_input() {
...
+    }
+
+    let total = allowed_count + filtered_count;
+    let filter_ratio = filtered_count as f64 / total as f64;
+    assert!(
+        filter_ratio >= 0.5,
+        "SC-002: at least 50% of tools should be filtered, got {:.0}%",
+        filter_ratio * 100.0
+    );
+}
```
[the remark this answers is not in the document]
Comment (brianluby) [reply crates/hooks/tests/e2e_test.rs:453]: Same rationale as above. The e2e test validates that the tool set composition achieves >=50% filtering (matching the real-world 182/352 ratio from the PRD). Actual filtering correctness is verified by the 13 unit tests in `tool_filter.rs` and the e2e test `post_tool_use_read_is_filtered_edit_is_stored` which confirms both allowed and disallowed tools exit successfully.

File: specs/013-memory-commit-optimization/prd.md [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+
+**Independent Test**: Store multiple observations and verify only one `commit()` occurs at flush time. Measure file size increase and confirm it is proportional to content volume.
+
+**Acceptance Scenarios**:
+1. **Given** a new session with 50 tool uses, **When** the session ends normally (stop hook fires), **Then** all observations are persisted in a single commit and the file size increase is proportional to content volume, not commit count.
+2. **Given** a session where `remember()` is called 10 times, **When** `flush()` is called once, **Then** all 10 observations are searchable and the backend received exactly one `commit()`.
+3. **Given** a session that crashes before the stop hook fires, **When** the next session starts, **Then** uncommitted observations from the crashed session are lost (acceptable trade-off), and the existing file remains uncorrupted.
+4. **Given** a long session where the buffer reaches 500 observations, **When** `remember()` is called with the 501st observation, **Then** an automatic flush commits the buffered 500 before buffering the new one.
```
[the remark this answers is not in the document]
Comment (brianluby) [reply specs/013-memory-commit-optimization/prd.md]: Fixed in 4f0cafa. Updated PRD acceptance scenario to match the implementation: auto-flush triggers when the buffer reaches 500 (on the 500th push), not on the 501st.

File: specs/013-memory-commit-optimization/prd.md [hunk 11]
```diff
@@ -0,0 +1,492 @@
...
+        usize count "Number of buffered observations"
+        usize byte_size "Approximate byte size of buffer"
+    }
+    DEDUP_CACHE {
+        HashSet hashes "Previously seen observation hashes"
+    }
+    MV2_FILE {
+        binary content "memvid-encoded observations"
+    }
+```
```
[the remark this answers is not in the document]
Comment (brianluby) [reply specs/013-memory-commit-optimization/prd.md:262]: Fixed in ab6a872. Removed phantom `Vec observations` from the ER diagram — the buffer only tracks `count` and `byte_size` metadata; observation data lives in the backend WAL after `put()`.

File: specs/013-memory-commit-optimization/prd.md [hunk 12]
```diff
@@ -0,0 +1,492 @@
...
+    /// No-op if buffer is empty. Acquires cross-process file lock.
+    pub fn flush(&mut self) -> Result<(), MindError>;
+}
+
+// Hook-level filtering (hooks crate)
+const ALLOWED_TOOLS: &[&str] = &["Edit", "Write", "NotebookEdit", "Bash"];
+
+/// Returns true if the tool name is in the allowed set for observation storage.
+fn is_tool_allowed(tool_name: &str) -> bool;
+```
```
[the remark this answers is not in the document]
Comment (brianluby) [reply specs/013-memory-commit-optimization/prd.md:286]: Fixed in ab6a872. Updated API signatures from `&mut self` to `&self` and corrected return types to match the actual implementation (interior mutability via Mutex).

**Comment by brianluby** [comment]
Regarding the `Default` impl nitpick for `ObservationBuffer` (from [review 3968522797](https://github.com/brianluby/rusty-brain/pull/47#pullrequestreview-3968522797)): Fixed in 84d70a3. Added `#[derive(Default)]` and updated `new()` to delegate to `Self::default()`.
